pkg/log/context.go

```diff
@@ -23,46 +23,66 @@ import (
 	"os"
 	"sync"
 
+	"github.com/onsi/ginkgo"
+
 	"github.com/onrik/logrus/filename"
+
 	"github.com/sirupsen/logrus"
 )
 
 type logKey struct{}
 
+const (
+	// FieldComponentName is the key of the component field in the log message.
+	FieldComponentName = "component"
+	// FieldCorrelationID is the key of the correlation id field in the log message.
+	FieldCorrelationID = "correlation_id"
+)
+
 var (
+	defaultEntry = logrus.NewEntry(logrus.StandardLogger())
+
 	supportedFormatters = map[string]logrus.Formatter{
 		"json": &logrus.JSONFormatter{},
 		"text": &logrus.TextFormatter{},
 	}
-	regMutex     = sync.Mutex{}
-	once         = sync.Once{}
-	defaultEntry = logrus.NewEntry(logrus.StandardLogger())
+
+	supportedOutputs = map[string]io.Writer{
+		os.Stdout.Name(): os.Stdout,
+		os.Stdin.Name():  os.Stdin,
+		os.Stderr.Name(): os.Stderr,
+		"ginkgowriter":   ginkgo.GinkgoWriter,
+	}
+	mutex           = sync.RWMutex{}
+	currentSettings = DefaultSettings()
+
 	// C is an alias for ForContext
 	C = ForContext
 	// D is an alias for Default
 	D = Default
 )
 
-const (
-	// FieldComponentName is the key of the component field in the log message.
-	FieldComponentName = "component"
-	// FieldCorrelationID is the key of the correlation id field in the log message.
-	FieldCorrelationID = "correlation_id"
-)
+func init() {
+	hook := filename.NewHook()
+	hook.Field = FieldComponentName
+	defaultEntry.Logger.AddHook(hook)
+	defaultEntry = defaultEntry.WithField(FieldCorrelationID, "-")
+	Configure(context.TODO(), currentSettings)
+}
 
 // Settings type to be loaded from the environment
 type Settings struct {
-	Level  string    `description:"minimum level for log messages"`
-	Format string    `description:"format of log messages. Allowed values - text, json"`
-	Output io.Writer `mapstructure:"-"`
+	Level  string `description:"minimum level for log messages" json:"level,omitempty"`
+	Format string `description:"format of log messages. Allowed values - text, json" json:"format,omitempty"`
+	Output string `description:"output for the logs. Allowed values - /dev/stdout, /dev/stdin, /dev/stderr, ginkgowriter" json:"output,omitempty"`
```

Comment:
This way you can change the output at runtime which may cause more bad than good IMO.

Reply:
check the previous comment